README
  LPC-Flash

  This program allow to read/write/erase/program memory (RAM and Flash)
for Philips LPC2xxx ARM microcontrollers over serial port.

Supported devices:
	LPC2103  LPC2104  LPC2105  LPC2106
	LPC2114  LPC2119  LPC2124  LPC2129
	LPC2131  LPC2132  LPC2134  LPC2136
	LPC2138  LPC2141  LPC2142  LPC2144
	LPC2146  LPC2148  LPC2194  LPC2210
	LPC2212  LPC2214  LPC2292  LPC2294
	LPC2364  LPC2366  LPC2368  LPC2378
	LPC2468

Author: Yuri Ovcharenko <amwsoft@gmail.com>


ChangeLog
2006-11-10  Yuri Ovcharenko

	* version 0.2.1.
	* More device support.
	
2006-11-09  Yuri Ovcharenko

	* version 0.2.0.
	* First public release.
	* Added UU encoding function.
	* Added write memory support.
	* Added go command.
	* Added flash programming support.

2006-11-07  Yuri Ovcharenko

	* version 0.1.0
	* Added read memory support.
	* Added UU decoding function.

2006-11-02  Yuri Ovcharenko

	* version 0.0.1
	* Project started.
	* Entering ISP mode by DTR/RTS.
	* Microcontroller detection.


src/uu.c
/***************************************************************************
 *   Copyright (C) 2006 by Yuri Ovcharenko                                 *
 *   amwsoft@gmail.com                                                     *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <sys/types.h>
#include <stdio.h>

#include "defs.h"
#include "uu.h"

extern int verbose;

int uudecode(char *in, char *out)
{
	int retval = 0;
	int i, j;
	char c;
	char buf[4];

	if (in[0] == 0x60)
		retval = 0;
	else
		retval = in[0] - 0x20;
	if (retval < 1)
		return 0;
	i = 1;
	j = 0;
	while (j < retval) {
		buf[0] = in[i + 0] == 0x60 ? 0 : in[i + 0] - 0x20;
		buf[1] = in[i + 1] == 0x60 ? 0 : in[i + 1] - 0x20;
		buf[2] = in[i + 2] == 0x60 ? 0 : in[i + 2] - 0x20;
		buf[3] = in[i + 3] == 0x60 ? 0 : in[i + 3] - 0x20;

		out[j + 0] = (buf[0] << 2) | ((buf[1] >> 4) & 0x3);
		out[j + 1] = (buf[1] << 4) | ((buf[2] >> 2) & 0xF);
		out[j + 2] = (buf[2] << 6) | ((buf[3]) & 0x3F);

		i += 4;
		j += 3;
	}
	VERBOSE(3, "UU decoded %d bytes.\n", retval);
	return retval;
}

int uuencode(unsigned char *in, size_t len, unsigned char *out)
{
	int retval;
	int i, j;
	u_int32_t buf;
	size_t count = len;

	if (!len)
		return 0;
	if (len > 45)
		return 0;

	while (count % 3) count++;
	out[0] = len;
	i = 0;
	j = 1;
	buf = 0;
	while (i < count) {
		buf <<= 8;
		if (i < len)
			buf |= in[i];
		if ((i % 3) == 2) {
			out[j++] = (buf >> 18) & 0x3F;
			out[j++] = (buf >> 12) & 0x3F;
			out[j++] = (buf >>  6) & 0x3F;
			out[j++] =  buf        & 0x3F;
			buf = 0;
		}
		i++;
	}
	retval = j;
	for (j = 0; j < retval; j++)
		out[j] = out[j] == 0 ? 0x60 : out[j] + 0x20;

	VERBOSE(3, "UU encoded %d bytes.\n", retval);
	return retval;
}

/* End of file */


src/uu.h
/***************************************************************************
 *   Copyright (C) 2006 by Yuri Ovcharenko                                 *
 *   amwsoft@gmail.com                                                     *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef _UU_H
#define _UU_H

int uudecode(char *in, char *out);
int uuencode(unsigned char *in, size_t len, unsigned char *out);

#endif /* __UU_H */


src/lpc.h
/***************************************************************************
 *   Copyright (C) 2006 by Yuri Ovcharenko                                 *
 *   amwsoft@gmail.com                                                     *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef _LPC_H
#define _LPC_H

#define MAX_IO_BUF	256
#define DEFAULT_TIMEOUT	500000	/* 0.5 sec */
#define DEFAULT_FREQ	18432	/* 18432000 Hz */

#define ERR_TIMEOUT		(ERROR_UC + 0x00)
#define ERR_ALIGN		(ERROR_UC + 0x01)
#define ERR_INCOMPLETE		(ERROR_UC + 0x02)
#define ERR_UNRECOGNIZED	(ERROR_UC + 0x03)
#define ERR_CONNECT		(ERROR_UC + 0x04)
#define ERR_READY		(ERROR_UC + 0x05)
#define ERR_COMMAND		(ERROR_UC + 0x06)

#define OP_READ			1
#define OP_WRITE		2
#define OP_ERASE		3
#define OP_FLASH		4

#define XON			0x11
#define XOFF			0x13

struct lpc_device {
	u_int32_t id;
	u_int32_t part;
	u_int32_t flash_size;
	u_int32_t ram_size;
	u_int32_t sectors;
	u_int32_t copy_size;
	u_int32_t *sector_map;
};

void ucprintdevs();
int ucisp(int freq);
int ucwaitresp(char *resp);
int ucwaitcmd(char *ibuf, size_t size, int *waitdata);
struct lpc_device *ucdetect(u_int32_t id);
int ucread(u_int32_t addr, char *buf, size_t len);
int ucwrite(u_int32_t addr, char *buf, size_t len, int level);
int ucerase(u_int32_t start_sector, u_int32_t sector_count);
int ucflash(u_int32_t addr, char *buf, size_t len);
int ucgo(u_int32_t addr, char mode);

#endif /* _LPC_H */


src/comm.c
/***************************************************************************
 *   Copyright (C) 2006 by Yuri Ovcharenko                                 *
 *   amwsoft@gmail.com                                                     *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <sys/types.h>
#include <stdio.h>
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <string.h>

#include "defs.h"
#include "comm.h"

extern int verbose;

static int port = -1;
static int baud = DEFAULT_BAUD;
static char *port_name = NULL;
static int oldmc = 0;

static struct termios ios, old_ios;

static char *set_port_name(char *name)
{
	if (port_name)
		free(port_name);
	if (name) {
		VERBOSE(3, "Set new port name\n");
		port_name = strdup(name);
	}
	else
		port_name = strdup(DEFAULT_PORT);
	return port_name;
}

int comm_open(char *name, int b)
{
	int mc;
	int retval;

	if (port > 0)
		comm_close();
	if (name)
		set_port_name(name);
	if (b) ios_init(b);

	VERBOSE(2, "Opening port...\n");
	port = open(port_name, O_RDWR | O_NONBLOCK | O_NOCTTY);
	if (port < 0) {
		perror("Can not open port");
		return -1;
	}
	VERBOSE(3, "Retrive old port attributes...\n");
	if (tcgetattr(port, &old_ios) < 0) {
		perror("Can not get port parameters");
		comm_close();
		return -1;
	}
	VERBOSE(3, "Apply new port attributes...\n");
	if (tcsetattr(port, TCSANOW, &ios) < 0) {
		perror("Can not set port parameters");
		comm_close();
		return -1;
	}

	VERBOSE(3, "Getting modem control lines state\n");
	retval = ioctl(port, TIOCMGET, &mc);
	if (retval < 0) {
		perror("Can not get modem control lines state");
		return retval;
	}
	oldmc = mc;
	mc &= (~(TIOCM_RTS | TIOCM_DTR));
	VERBOSE(3, "Setting RTS & DTR lines to OFF\n");
	retval = ioctl(port, TIOCMSET, &mc);
	if (retval < 0) {
		perror("Can not set RTS & DTR lines");
		return retval;
	}

	VERBOSE(2, "Opened.\n");
	return 0;
}

void comm_close()
{
	int retval;

	VERBOSE(2, "Closing port...\n");
	if (port < 0)
		return;
	VERBOSE(3, "Apply old modem control lines state...\n");
	retval = ioctl(port, TIOCMSET, &oldmc);
	if (retval < 0) {
		perror("Can not set RTS & DTR lines");
		return retval;
	}
	VERBOSE(3, "Apply old port attributes...\n");
	tcsetattr(port, TCSANOW, &old_ios);
	close(port);
	port = -1;
}

int comm_opened()
{
	return port >= 0;
}

static void ios_init(int b)
{
	VERBOSE(2, "Initializing termios structure.\n");
	cfmakeraw(&ios);

	/* Input flags */
	ios.c_iflag |= IGNBRK | IGNPAR;
	/* Control flags */
	ios.c_cflag |= CREAD | CLOCAL;
	ios.c_cflag &= ~CRTSCTS;
	/* Control char's */
	ios.c_cc[VINTR] = 0;
	ios.c_cc[VQUIT] = 0;
	ios.c_cc[VERASE] = 0;
	ios.c_cc[VKILL] = 0;
	ios.c_cc[VEOF] = 0;
	ios.c_cc[VMIN] = 0;
	ios.c_cc[VEOL] = 0;
	ios.c_cc[VTIME] = 0;
	ios.c_cc[VEOL2] = 0;
	ios.c_cc[VSTART] = 0;
	ios.c_cc[VSTOP] = 0;
	ios.c_cc[VSUSP] = 0;
	ios.c_cc[VLNEXT] = 0;
	ios.c_cc[VWERASE] = 0;
	ios.c_cc[VREPRINT] = 0;
	ios.c_cc[VDISCARD] = 0;

	baud = DEFAULT_BAUD;
	switch (b) {
		case 9600:
			baud = 9600;
			break;
		case 19200:
			baud = 19200;
			break;
		case 38400:
			baud = 38400;
			break;
		case 57600:
			baud = B57600;
			break;
		case 115200:
			baud = B115200;
			break;
		case 230400:
			baud = B230400;
			break;
	}
	cfsetospeed(&ios, baud);
	cfsetispeed(&ios, baud);
}

void comm_init(char *name, int baud)
{
	set_port_name(name);
	ios_init(baud);
}

void comm_clean()
{
	comm_close();
	FREE(port_name);
}

void comm_info()
{
	printf("Port:\t\t%s %s\n", port_name, comm_opened() ? "opened." : "closed.");
	printf("Baud:\t\t");
	switch (baud) {
		case B9600:
			printf("9600");
			break;
		case B19200:
			printf("19200");
			break;
		case B38400:
			printf("38400");
			break;
		case B115200:
			printf("115200");
			break;
		case B230400:
			printf("230400");
			break;
		case B57600:
		default:
			printf("57600");
			break;
	}
	printf(".\n");
}

int comm_read(char *data, size_t size)
{
	int retval = 0;

	if (port < 0)
		retval = comm_open(NULL, 0);
	if (retval < 0)
		return retval;
	VERBOSE(4, "Reading from device...\n");
	return read(port, data, size);
}

int comm_write(char *data, size_t len)
{
	int retval = 0;
	int count;

	if (port < 0)
		retval = comm_open(NULL, 0);
	if (retval < 0)
		return retval;

	count = len;
	while (count > 0) {
		VERBOSE(4, "Writing to device... ");
		retval = write(port, data, count);
		if (retval < 0) {
			if (errno == EAGAIN)
				continue;
			else
				return retval;
		}
		VERBOSE(4, "%d bytes.\n", retval);
		count -= retval;
		data += retval;
	}

	return len;
}

int comm_setrts(int rts)
{
	int retval = 0;
	int mc;

	if (port < 0)
		retval = comm_open(NULL, 0);
	if (retval < 0)
		return retval;

	VERBOSE(3, "Getting modem control lines state\n");
	retval = ioctl(port, TIOCMGET, &mc);
	if (retval < 0) {
		perror("Can not get modem control lines state");
		return retval;
	}
	VERBOSE(4, "mc = 0x%08X TIOCM_RTS = 0x%08X\n", mc, TIOCM_RTS);
	if (rts)
		mc |= TIOCM_RTS;
	else
		mc &= (~TIOCM_RTS);
	VERBOSE(4, "mc = 0x%08X TIOCM_RTS = 0x%08X\n", mc, TIOCM_RTS);
	VERBOSE(3, "Setting RTS line to %s\n", rts ? "ON" : "OFF");
	retval = ioctl(port, TIOCMSET, &mc);
	if (retval < 0) {
		perror("Can not set RTS line");
		return retval;
	}
	return 0;
}

int comm_setdtr(int dtr)
{
	int retval = 0;
	int mc;

	if (port < 0)
		retval = comm_open(NULL, 0);
	if (retval < 0)
		return retval;

	VERBOSE(3, "Getting modem control lines state\n");
	retval = ioctl(port, TIOCMGET, &mc);
	if (retval < 0) {
		perror("Can not get modem control lines state");
		return retval;
	}
	VERBOSE(4, "mc = 0x%08X TIOCM_DTR = 0x%08X\n", mc, TIOCM_DTR);
	if (dtr)
		mc |= TIOCM_DTR;
	else
		mc &= (~TIOCM_DTR);
	VERBOSE(4, "mc = 0x%08X TIOCM_DTR = 0x%08X\n", mc, TIOCM_DTR);
	VERBOSE(3, "Setting DTR line to %s\n", dtr ? "ON" : "OFF");
	retval = ioctl(port, TIOCMSET, &mc);
	if (retval < 0) {
		perror("Can not set DTR line");
		return retval;
	}
	return 0;
}

/* End of file */


src/comm.h
/***************************************************************************
 *   Copyright (C) 2006 by Yuri Ovcharenko                                 *
 *   amwsoft@gmail.com                                                     *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef _COMM_H
#define _COMM_H

#include <termios.h>

#define DEFAULT_PORT	"/dev/tts/0"
#define DEFAULT_BAUD	B57600;

void comm_init(char *name, int baud);
void comm_clean();
int comm_open(char *name, int b);
void comm_close();
int comm_opened();
int comm_read(char *data, size_t len);
int comm_write(char *data, size_t len);
void comm_info();

static char *set_port_name(char *name);
static void ios_init(int b);

int comm_setrts(int rts);
int comm_setdtr(int dtr);

#endif /* _COMM_H */


src/defs.h
/***************************************************************************
 *   Copyright (C) 2006 by Yuri Ovcharenko                                 *
 *   amwsoft@gmail.com                                                     *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef _DEFS_H
#define _DEFS_H

#define PROJECT		"LPC-Flash"
#define PROJECT_VERSION	"0.2.4"

#define VERBOSE(level, fmt, arg...)			\
		{					\
			if (level <= verbose) {		\
				printf(fmt, ## arg);	\
				fflush(stdout);		\
			}				\
		}

#define ERROR_CMD	0x00
#define ERROR_UC	0x10
#define ERROR_IO	0x20
#define ERROR_BATCH	0x30

#define FREE(x)         if (x != NULL) {free(x); x = NULL;}

#endif /* _DEFS_H */


src/lpcflash.c
/***************************************************************************
 *   Copyright (C) 2006 by Yuri Ovcharenko                                 *
 *   amwsoft@gmail.com                                                     *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>

#include "defs.h"
#include "comm.h"
#include "lpc.h"

int verbose = 0;
int freq = DEFAULT_FREQ;

struct option long_opt [] = {
	{"interface", required_argument, 0, 'i'},
	{"baud", required_argument, 0, 'b'},
	{"freq", required_argument, 0, 'f'},

	{"read", required_argument, 0, 'r'},
	{"write", required_argument, 0, 'w'},
	{"erase", required_argument, 0, 'e'},
	{"programm", required_argument, 0, 'p'},

	{"go", required_argument, 0, 'g'},

	{"file", required_argument, 0, 'l'},

	{"verbose", no_argument, 0, 'v'},
	{"quiet", no_argument, 0, 'q'},
	{"version", no_argument, 0, 'V'},
	{"help", no_argument, 0, 'h'},
	{0, 0, 0, 0},
};

void usage();

int main(int argc, char *argv[])
{
	int retval;
	int i;
	int opt;
	char *c = NULL;
	int opt_index;
	char *port = NULL;
	int baud = 0;
	char *ok;
	char *cmd = NULL;
	int operation = 0;
	u_int32_t addr;
	u_int32_t goaddr;
	char gomode;
	int goflag = 0;
	size_t size, msize;
	char *filename = NULL;
	int fd;
	size_t pos;
	size_t count;
	int n;
	struct stat st;

	while (1) {
		opt = getopt_long(argc, argv, "i:b:f:r:w:e:p:g:l:vqVh", long_opt, &opt_index);
		if (opt == -1)
			break;
		switch (opt) {
			case 0:
				printf("\ngetopt_long == 0\noption == '%s'\n", long_opt[opt_index].name);
				if (optarg)
					printf(" with arg '%s'", optarg);
				printf("\n");
				break;
			case 'i':
				port = optarg;
				break;
			case 'b':
				baud = strtol(optarg, &ok, 0);
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid baud rate: '%s'\n", optarg);
					return 254;
				}
				break;
			case 'f':
				freq = strtol(optarg, &ok, 0);
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid frequency: '%s'\n", optarg);
					return 254;
				}
				if ((freq < 10000) || (freq > 100000)){
					fprintf(stderr, "Invalid frequency: '%d kHz'\n", freq);
					return 254;
				}
				break;
			case 'r':
				if (operation) {
					fprintf(stderr, "Only one operation may be specified.\n");
					return 254;
				}
				c = strstr(optarg, ":");
				if (c == NULL) {
					fprintf(stderr, "Invalid argument format for -%c option.\n", opt);
					return 254;
				}
				*c = 0;
				c++;
				addr = strtoul(optarg, &ok, 0);
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid address: '%s'\n", optarg);
					return 254;
				}
				if (addr & 3) {
					fprintf(stderr, "Invalid address: '%s'\n", optarg);
					return 254;
				}
				size = strtoul(c, &ok, 0);
				c--;
				*c = ':';
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid size: '%s'\n", optarg);
					return 254;
				}
				if (size == 0) {
					fprintf(stderr, "ZERO size.\n");
					return 254;
				}
				if (size & 3) {
					fprintf(stderr, "Invalid size: '%s'\n", optarg);
					return 254;
				}
				VERBOSE(2, "Operation READ 0x%08X:%d.\n", addr, size);
				operation = OP_READ;
				break;
			case 'w':
				if (operation) {
					fprintf(stderr, "Only one operation may be specified.\n");
					return 254;
				}
				addr = strtoul(optarg, &ok, 0);
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid address: '%s'\n", optarg);
					return 254;
				}
				if (addr & 3) {
					fprintf(stderr, "Invalid address: '%s'\n", optarg);
					return 254;
				}
				VERBOSE(2, "Operation WRITE 0x%08X.\n", addr);
				operation = OP_WRITE;
				break;
			case 'e':
				if (operation) {
					fprintf(stderr, "Only one operation may be specified.\n");
					return 254;
				}
				c = strstr(optarg, ":");
				if (c == NULL) {
					fprintf(stderr, "Invalid argument format for -%c option.\n", opt);
					return 254;
				}
				*c = 0;
				addr = strtoul(optarg, &ok, 0);
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid sector number: '%s'\n", optarg);
					return 254;
				}
				*c++ = ':';
				size = strtoul(c, &ok, 0);
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid setor count: '%s'\n", c);
					return 254;
				}
				if (size == 0) {
					fprintf(stderr, "ZERO sector count.\n");
					return 254;
				}
				VERBOSE(2, "Operation ERASE %d:%d.\n", addr, size);
				operation = OP_ERASE;
				break;
			case 'p':
				if (operation) {
					fprintf(stderr, "Only one operation may be specified.\n");
					return 254;
				}
				addr = strtoul(optarg, &ok, 0);
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid address: '%s'\n", optarg);
					return 254;
				}
				if (addr & 0x1FF) {
					fprintf(stderr, "Invalid address: '%s'\n", optarg);
					return 254;
				}
				VERBOSE(2, "Operation PROGRAM 0x%08X.\n", addr);
				operation = OP_FLASH;
				break;
			case 'g':
				c = strstr(optarg, ":");
				if (c == NULL) {
					fprintf(stderr, "Invalid argument format for -%c option.\n", opt);
					return 254;
				}
				*c = 0;
				c++;
				goaddr = strtoul(optarg, &ok, 0);
				if (!(*ok == '\0')) {
					fprintf(stderr, "Invalid address for -%c option: '%s'\n", opt, optarg);
					return 254;
				}
				if (goaddr & 3) {
					fprintf(stderr, "Invalid address for -%c option: '%s'\n", opt, optarg);
					return 254;
				}
				gomode = *c;
				c--;
				*c = ':';
				switch (gomode) {
					case 'T':
					case 't':
						gomode = 'T';
						break;
					case 'A':
					case 'a':
						gomode = 'A';
						break;
					default:
						fprintf(stderr, "Invalid mode for -%c option: '%s'\n", opt, optarg);
						return 254;
						break;
				}
				goflag = 1;
				break;
			case 'l':
				filename = optarg;
				break;
			case 'v':
				++verbose;
				break;
			case 'q':
				verbose = -1;
				break;
			case 'V':
				printf(PROJECT " version " PROJECT_VERSION "\n");
				return 0;
				break;
			case 'h':
				usage();
				return 0;
				break;
			case '?':
				return 254;
				break;
		}
	}

	c = NULL;

	comm_init(port, baud);

	retval = ucisp(freq);
	if (retval) {
		comm_clean();
		return 2;
	}

	switch (operation) {
		/***************************************************/
		case OP_READ:
			c = malloc(size);
			if (c == NULL) {
				perror("Can not allocate memory");
			} else {
				retval = ucread(addr, c, size);
				if (!retval) {
					if (filename != NULL) {
						fd = open(filename, O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
						if (fd < 0) {
							perror("Can not create file");
						} else {
							retval = write(fd, c, size);
							if (retval < 0) {
								perror("Can not write data to file");
							}
							close(fd);
						}
					} else {
						for (i = 0; i < (size >> 2); i++) {
							printf("0x%08X\n", ((u_int32_t *)c)[i]);
						}
					}
				}
			}
			break;
		/***************************************************/
		case OP_WRITE:
			if (filename == NULL) {
				fprintf(stderr, "File to write is not specified.\n");
				return 254;
			}
			retval = stat(filename, &st);
			if (retval < 0) {
				perror("Can not stat file");
				return 254;
			}
			if (!S_ISREG(st.st_mode)) {
				fprintf(stderr, "%s is not a regular file.\n", filename);
				return 254;
			}
			size = st.st_size;
			if (size == 0) {
				fprintf(stderr, "Zero file size.\n");
				return 254;
			}
			while (size & 3)
				size++;
			fd = open(filename, O_RDONLY);
			if (fd < 0) {
				perror("Can not open file");
				return 254;
			}

			c = malloc(size);
			if (c == NULL) {
				perror("Can not allocate memory");
				return 254;
			} else {
				memset(c, 0, size);
				retval = read(fd, c, st.st_size);
				if (retval != st.st_size) {
					perror("Can not read data");
					close(fd);
				} else {
					close(fd);
					retval = ucwrite(addr, c, size, 0);
					if (retval) {
						fprintf(stderr, "Error writing to microcontroller.\n");
					}
				}
				FREE(c);
			}
			break;
		/***************************************************/
		case OP_ERASE:
			retval = ucerase(addr, size);
			if (retval) {
				fprintf(stderr, "Error erasing flash.\n");
			}
			break;
		/***************************************************/
		case OP_FLASH:
			if (filename == NULL) {
				fprintf(stderr, "File to program is not specified.\n");
				return 254;
			}
			retval = stat(filename, &st);
			if (retval < 0) {
				perror("Can not stat file");
				return 254;
			}
			if (!S_ISREG(st.st_mode)) {
				fprintf(stderr, "%s is not a regular file.\n", filename);
				return 254;
			}
			size = st.st_size;
			if (size == 0) {
				fprintf(stderr, "Zero file size.\n");
				return 254;
			}
			fd = open(filename, O_RDONLY);
			if (fd < 0) {
				perror("Can not open file");
				return 254;
			}

			msize = size;
			if (size & 3UL) {
				msize &= ~3UL;
				msize += 4;
			}
			c = malloc(msize);
			if (c == NULL) {
				perror("Can not allocate memory");
				return 254;
			} else {
				memset(c, 0xFF, size);
				retval = read(fd, c, size);
				if (retval != size) {
					perror("Can not read data");
					close(fd);
				} else {
					close(fd);
					retval = ucflash(addr, c, msize);
					if (retval) {
						fprintf(stderr, "Error programming.\n");
					}
				}
				FREE(c);
			}
			break;
	}

	if ((!retval) && (goflag)) {
		retval = ucgo(goaddr, gomode);
	}

	FREE(c);

	comm_clean();

	return retval;
}

void usage()
{
	/*          0         1         2         3         4         5         6         7       */
	/*      01234567890123456789012345678901234567890123456789012345678901234567890123456789  */
	printf("Usage: lpcflash [OPTIONS]\n");
	printf("OPTIONS:\n");
	printf("\n");

	printf("\t-i DEVICE\n\t--interface=DEVICE\n");
	printf("\t\tUse DEVICE to communicate with microcontroller (" DEFAULT_PORT ").\n");
	printf("\n");

	printf("\t-b BAUDRATE\n\t--baud=BAUDRATE\n");
	printf("\t\tCommunicate with microcontroller at the given BAUDRATE (57600).\n");
	printf("\n");

	printf("\t-f FREQUENCY\n\t--freq=FREQUENCY\n");
	printf("\t\tMicrocontroller running at FREQUENCY kHz (18432).\n");
	printf("\n");

	printf("\t-r ADDRESS:SIZE\n\t--read=ADDRESS:SIZE\n");
	printf("\t\tRead from memory SIZE bytes at the given ADDRESS.\n");
	printf("\t\tThe address must be word-aligned. The size must be multiple of 4.\n");
	printf("\t\tData stored to file specified with -l option.\n");
	printf("\n");

	printf("\t-w ADDRESS\n\t--write=ADDRESS\n");
	printf("\t\tWrite to RAM at the given ADDRESS.\n");
	printf("\t\tThe address must be word-aligned.\n");
	printf("\t\tData loaded from file specified with -l option.\n");
	printf("\n");

	printf("\t-e START:COUNT\n\t--erase=START:COUNT\n");
	printf("\t\tErase COUNT flash sectors begining from sector number START.\n");
	printf("\n");

	printf("\t-p ADDRESS\n\t--program=ADDRESS\n");
	printf("\t\tProgram flash begining at the given ADDRESS.\n");
	printf("\t\tThe address must be 512byte-aligned.\n");
	printf("\t\tData loaded from file specified with -l option.\n");
	printf("\t\tNOTE:\n");
	printf("\t\tThe boot signature at address 0x00000014 is calculated if and\n");
	printf("\t\tonly if ADDRESS = 0 and data size to be programmed is greater\n");
	printf("\t\tthen or equal 32 bytes.\n");
	printf("\t\tWhile programming less then 32 bytes at ADDRESS = 0 the\n");
	printf("\t\tresult is unpredictable.\n");
	printf("\t\tWARNING:\n");
	printf("\t\tThe read/modify/write is not supported!\n");
	printf("\n");

	printf("\t-g ADDRESS:MODE\n\t--go=ADDRESS:MODE\n");
	printf("\t\tRun user program at the given ADDRESS in specified MODE.\n");
	printf("\t\tMODE can be one of T or A character.\n");
	printf("\t\tT is for Thumb and A is for ARM mode execution.\n");
	printf("\t\tLowercase characters is allowed.\n");
	printf("\t\tThe address must be word-aligned.\n");
	printf("\t\tThe go comman is executed after successful execution of all\n");
	printf("\t\tother commands.\n");
	printf("\n");

	printf("\t-l FILENAME\n\t--file=FILENAME\n");
	printf("\t\tThe means of this option depends on operation.\n");
	printf("\t\tFor read operation data stored in the file with name FILENAME.\n");
	printf("\t\tFor write and program operations data loaded from file with\n");
	printf("\t\tname FILENAME.\n");
	printf("\n");

	printf("\t-v\n\t--verbose\n");
	printf("\t\tBe verbose. More options increase verbose level.\n");
	printf("\n");

	printf("\t-q\n\t--quiet\n");
	printf("\t\tBe quiet. No output to stdout.\n");
	printf("\n");

	printf("\t-V\n\t--version\n");
	printf("\t\tPrint version and exit.\n");
	printf("\n");

	printf("\t-h\n\t--help\n");
	printf("\t\tPrint this screan and exit.\n");
	printf("\n");

	ucprintdevs();
}

/* End of file */
